File: javascript/src/providers/Mailnesia.js
import { TempMailProvider } from '../base.js';
import { Message, MessageDetail } from '../models.js';
import { TempMailError } from '../errors.js';

const BASE = 'https://mailnesia.com';
const CHARS = 'abcdefghijklmnopqrstuvwxyz0123456789';

/**
 * mailnesia.com provider — public mailbox, no auth, HTML scraping.
 *
 * Inbox is a `<table>`; each row is sender / subject / time with a link whose
 * trailing path segment is the message id. Read view exposes the body inside
 * `<div id="message">`.
 */
export class Mailnesia extends TempMailProvider {
  /** @param {{ randomUA?: boolean, useCookies?: boolean }|undefined} opts */
  constructor(opts = {}) {
    super('mailnesia', { randomUA: opts.randomUA, useCookies: true });
    /** @type {string|null} */
    this._email = null;
    /** @type {string|null} */
    this._username = null;
  }

  /** @returns {Promise<string>} */
  async generateEmail() {
    this._username = Array.from({ length: 10 }, () =>
      CHARS[Math.floor(Math.random() * CHARS.length)],
    ).join('');
    this._email = `${this._username}@mailnesia.com`;
    return this._email;
  }

  /**
   * Generate random IP for header rotation
   * @returns {string}
   */
  _generateRandomIP() {
    return `${Math.floor(Math.random() * 254) + 1}.${Math.floor(Math.random() * 256)}.${Math.floor(Math.random() * 256)}.${Math.floor(Math.random() * 254) + 1}`;
  }

  /**
   * Get headers with rotated IP to bypass 403
   * @returns {Object}
   */
  _getHeadersWithIPRotation() {
    const ip = this._generateRandomIP();
    return {
      'X-Forwarded-For': ip,
      'X-Real-IP': ip,
      'CF-Connecting-IP': ip,
      'True-Client-IP': ip,
    };
  }

  /**
   * @param {string} email
   * @returns {Promise<Message[]>}
   */
  async getInbox(email) {
    const username = email.includes('@') ? email.split('@')[0] : email;
    const headers = this._getHeadersWithIPRotation();
    const res = await this._fetch(`${BASE}/mailbox/${username}`, { headers });
    const html = await res.text();
    const messages = [];
    const rowRe = /<tr[^>]*>([\s\S]*?)<\/tr>/gi;
    let rm;
    while ((rm = rowRe.exec(html))) {
      const row = rm[1];
      const cells = [];
      const cellRe = /<td[^>]*>([\s\S]*?)<\/td>/gi;
      let cm;
      while ((cm = cellRe.exec(row))) cells.push(cm[1]);
      if (cells.length < 3) continue;
      const sender = this._stripHtml(cells[0]);
      const subject = this._stripHtml(cells[1]);
      const time = this._stripHtml(cells[2]);
      const href = (row.match(/<a[^>]*href="([^"]+)"/) || [])[1] || '';
      const msgId = href ? href.replace(/\/+$/, '').split('/').pop() : '';
      if (sender || subject) {
        messages.push(new Message({ id: msgId, sender, subject, date: time }));
      }
    }
    return messages;
  }

  /**
   * @param {string} messageId
   * @returns {Promise<MessageDetail>}
   */
  async readMessage(messageId) {
    if (!this._username) {
      throw new TempMailError('No email — call generateEmail() first', this.name);
    }
    const headers = this._getHeadersWithIPRotation();
    const res = await this._fetch(`${BASE}/mailbox/${this._username}/${messageId}`, { headers });
    const html = await res.text();
    const body = this._extractBlockById(html, 'message');
    const bodyHtml = body || html;
    const bodyText = body ? this._stripHtml(body) : this._stripHtml(html);
    return new MessageDetail({
      id: messageId,
      sender: '',
      subject: '',
      date: new Date(),
      bodyText,
      bodyHtml,
      attachments: [],
    });
  }

  /**
   * @param {string} _email
   * @returns {Promise<boolean>}
   */
  async deleteEmail(_email) {
    this._email = null;
    this._username = null;
    return true;
  }

  // --- regex HTML helpers (no DOMParser in Node, no deps) ---

  /**
   * Extract the inner HTML of `<div id="{id}">` by counting balanced div tags.
   * @param {string} html
   * @param {string} id
   * @returns {string|null}
   */
  _extractBlockById(html, id) {
    const open = html.match(new RegExp(`<div[^>]*id="${id}"[^>]*>`, 'i'));
    if (!open) return null;
    const start = open.index + open[0].length;
    let depth = 1;
    const tagRe = /<\/?div\b[^>]*>/gi;
    tagRe.lastIndex = start;
    let t;
    while ((t = tagRe.exec(html))) {
      depth += t[0].startsWith('</') ? -1 : 1;
      if (depth === 0) return html.slice(start, t.index);
    }
    return html.slice(start);
  }

  /**
   * Strip HTML tags and decode common entities.
   * @param {string} html
   * @returns {string}
   */
  _stripHtml(html) {
    if (!html) return '';
    return html
      .replace(/<[^>]*>/g, '')
      .replace(/&nbsp;/g, ' ')
      .replace(/&amp;/g, '&')
      .replace(/&lt;/g, '<')
      .replace(/&gt;/g, '>')
      .replace(/&quot;/g, '"')
      .trim();
  }
}

File: go/providers/mailnesia.go
package providers

import (
	"fmt"
	"math/rand"
	"regexp"
	"strings"
	"time"
)

const mailnesiaBaseURL = "https://mailnesia.com"

// Mailnesia is the mailnesia.com provider (public mailbox, HTML scraping with IP rotation).
type Mailnesia struct {
	client   *HttpClient
	email    string
	username string
}

// NewMailnesia creates a new Mailnesia provider.
func NewMailnesia(config map[string]string) *Mailnesia {
	proxies := ParseProxies(config["proxies"])
	randomUA := config["randomUa"] == "true"
	return &Mailnesia{client: NewHttpClient(proxies, randomUA, true)}
}

func (m *Mailnesia) generateRandomIP() string {
	return fmt.Sprintf("%d.%d.%d.%d",
		rand.Intn(254)+1,
		rand.Intn(256),
		rand.Intn(256),
		rand.Intn(254)+1)
}

func (m *Mailnesia) getHeadersWithIPRotation() map[string]string {
	ip := m.generateRandomIP()
	return map[string]string{
		"X-Forwarded-For":  ip,
		"X-Real-IP":        ip,
		"CF-Connecting-IP": ip,
		"True-Client-IP":   ip,
	}
}

func (m *Mailnesia) GenerateEmail() (string, error) {
	m.username = randName(10)
	m.email = m.username + "@mailnesia.com"
	return m.email, nil
}

func (m *Mailnesia) GetInbox(email string) ([]Message, error) {
	username := email
	if i := strings.Index(email, "@"); i > 0 {
		username = email[:i]
	}
	headers := m.getHeadersWithIPRotation()
	body, err := m.client.GetWithHeaders(mailnesiaBaseURL+"/mailbox/"+username, headers)
	if err != nil {
		return nil, &TempMailError{Provider: "mailnesia", Message: err.Error()}
	}
	html := string(body)
	var msgs []Message
	for _, row := range mailnesiaRowRe.FindAllStringSubmatch(html, -1) {
		cells := mailnesiaCellRe.FindAllStringSubmatch(row[1], -1)
		if len(cells) < 3 {
			continue
		}
		sender := stripHTML(cells[0][1])
		subject := stripHTML(cells[1][1])
		timeStr := stripHTML(cells[2][1])
		if sender == "" && subject == "" {
			continue
		}
		msgID := ""
		if lm := mailnesiaLinkRe.FindStringSubmatch(row[1]); len(lm) > 1 {
			msgID = lastPathSegment(lm[1])
		}
		msgs = append(msgs, Message{ID: msgID, Sender: sender, Subject: subject, Date: parseMailDate(timeStr)})
	}
	return msgs, nil
}

func (m *Mailnesia) ReadMessage(messageID string) (*MessageDetail, error) {
	if m.username == "" {
		return nil, &TempMailError{Provider: "mailnesia", Message: "no email — call GenerateEmail() first"}
	}
	headers := m.getHeadersWithIPRotation()
	body, err := m.client.GetWithHeaders(mailnesiaBaseURL+"/mailbox/"+m.username+"/"+messageID, headers)
	if err != nil {
		return nil, &TempMailError{Provider: "mailnesia", Message: err.Error()}
	}
	html := string(body)
	bodyHTML := extractDivContent(html, "message")
	bodyText := stripHTML(bodyHTML)
	if bodyHTML == "" {
		bodyText = stripHTML(html)
	}
	return &MessageDetail{
		Message:  Message{ID: messageID, Date: time.Now()},
		BodyText: bodyText,
		BodyHTML: bodyHTML,
	}, nil
}

func (m *Mailnesia) DeleteEmail(email string) (bool, error) {
	m.email = ""
	m.username = ""
	return true, nil
}

func (m *Mailnesia) WaitForEmail(email string, timeout time.Duration, interval time.Duration) (*Message, error) {
	return waitForEmail(m, email, timeout, interval)
}

var (
	mailnesiaRowRe  = regexp.MustCompile(`(?s)<tr[^>]*>(.*?)</tr>`)
	mailnesiaCellRe = regexp.MustCompile(`(?s)<td[^>]*>(.*?)</td>`)
	mailnesiaLinkRe = regexp.MustCompile(`href="([^"]*)"`)
)
